server/src/main/java/org/elasticsearch/common/util/Maps.java

```diff
@@ -81,12 +81,27 @@
      * @param <V>   the type of the values in the map
      * @return an immutable map that contains the items from the specified map with the provided key removed
      */
+    @SuppressWarnings("unchecked")
     public static <K, V> Map<K, V> copyMapWithRemovedEntry(final Map<K, V> map, final K key) {
         Objects.requireNonNull(map);
         Objects.requireNonNull(key);
         assert checkIsImmutableMap(map, key, map.get(key));
-        return map.entrySet().stream().filter(k -> key.equals(k.getKey()) == false)
-            .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
+        if (map.containsKey(key) == false) {
+            return map;
+        }
+        final int size = map.size();
+        if (size == 1) {
+            return Map.of();
+        }
+        @SuppressWarnings("rawtypes")
+        final Map.Entry<K, V>[] entries = new Map.Entry[size - 1];
+        int i = 0;
+        for (Map.Entry<K, V> entry : map.entrySet()) {
+            if (key.equals(entry.getKey()) == false) {
+                entries[i++] = entry;
+            }
+        }
+        return Map.ofEntries(entries);
```

Comment:
This feels a lot like premature optimization, is this really a bottleneck here?

Reply:
This whole thing showed up hot during cluster restart + reroute tests on large shard count benchmarks. The problem with those really is that all these `stream` things look nice and their overhead might be irrelevant most of the time, now show up all over the place when they get nested inside other loops and we simply only have a single master update thread :)
I think the 7.x code is different for the data tier allocation decider that's why it shows up there, for 8.x this is less of a relevant change probably but as you point out will help with the filter decider.